### ISSUE TITLE
Improve API for snippet reuse

### ISSUE THREAD

**Issue by pcornelissen** [spring-projects/spring-restdocs#844]
The current API already has in parts support to reduce duplication, but there are cases, which are not covered yet.

For example you have a "page" result that is returned as single entity and in lists. When this page is more complex, for example like this:

```java
    public static final ResponseFieldsSnippet PAGE_RESPONSE_DEFINITION = responseFields(
            PAGE_ID_FIELD_DESCRIPTORS
    ).and(
            PAGE_COMMON_FIELD_DESCRIPTORS
    ).andWithPrefix("auditMetaData.",
            AUDITMETADATA_FIELD_DESCRIPTORS
    ).and(
            PAGE_CONTENT_FIELD_DESCRIPTORS
    ).and(
            LINKS_FIELD_DESCRIPTOR
    );
```
then it's not really feasable to have one large list of descriptors when you define the fields.

But then you have a collection that is exposed using the HATEOAS collections, so the result is still a pae response, but it's "encapsulated" in something like ```_embedded.pages[].*```.

As workaround, I have created a snippet like that:

```java
    public static final ResponseFieldsSnippet PAGES_RESPONSE_DEFINITION =
            responseFields(fieldWithPath("_embedded.pages").description("List of pages matching the query. If the result is empty, even the _embedded path will not be there"))
                    .andWithPrefix("_embedded.pages[]",
                            PAGE_ID_FIELD_DESCRIPTORS)
                    .andWithPrefix("_embedded.pages[]",
                            PAGE_COMMON_FIELD_DESCRIPTORS)
                    .andWithPrefix("_embedded.pages[].auditMetaData.",
                            AUDITMETADATA_FIELD_DESCRIPTORS)
                    .andWithPrefix("_embedded.pages[]",
                            PAGE_CONTENT_FIELD_DESCRIPTORS)
                    .andWithPrefix("_embedded.pages[]",
                            LINKS_FIELD_DESCRIPTOR);
```

There is still plenty of duplication, that I don't know how to get rid of right now. 

I have an idea, but the ```ResponseFieldsSnippet``` does not expose the list of descriptors. If that would be the case, I could just do something like this:

```java
    public static final ResponseFieldsSnippet PAGES_RESPONSE_DEFINITION =
            responseFields(fieldWithPath("_embedded.pages").description("List of pages matching the query. If the result is empty, even the _embedded path will not be there"))
                    .andWithPrefix("_embedded.pages[]",
                            PAGE_RESPONSE_DEFINITION.getDescriptors())
```

Of course there are also other ways, which would be even better like

```java
    public static final ResponseFieldsSnippet PAGES_RESPONSE_DEFINITION =
            responseFields(fieldWithPath("_embedded.pages").description("List of pages matching the query. If the result is empty, even the _embedded path will not be there"))
                    .andEmbeddWithPrefix("_embedded.pages[]",
                            PAGE_RESPONSE_DEFINITION)
```

I assume that is something that many people face. Maybe I have overlooked something in the API, in that case I'd gladly close this and refactor my code. :-)

**Comment by wilkinsona** [comment]
Thanks for raising this. Thus far, what I have recommended is that you avoid the duplication by cross-linking in the documentation. You can see an [example of this](https://docs.spring.io/spring-restdocs/docs/2.0.x/samples/restful-notes/api-guide.html#resources_notes_list_response_fields) in the sample RESTful notes API guide. Here `_embedded.notes` is documented as being an array of `Note` resources, with a cross-link to [its documentation](https://docs.spring.io/spring-restdocs/docs/2.0.x/samples/restful-notes/api-guide.html#resources_note). In a HATEOAS-based API where there may be a lot of cross-referencing and embedded resources, I think this approach will produce documentation that is more concise and easier for the reader to consume.

That said, I can see why you may have a preference for precisely documenting each resource, including its embedded resources so I think some enhancements in this area are still worth considering. I would, however, be interested to know if you've considered the approach above and if it would meet your needs.

**Comment by spring-projects-issues** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by pcornelissen** [comment]
Thanks @wilkinsona for the hint. I haven't notice the description "autolinking" and I just checked, I couldn't see it in the documentation, It appears to be only in this example.

Maybe the Docs should be clarified that this feature exists and how it basically works. This might redice the chance for similar questions in the future.

As you said, the link approach is ok for now. I just have to figure out how to use it properly. But it's not as self contained as it could be. So having a feature as I suggested would be a good addition.
